[48] Update plugin for 4.3

JustinKyleJames opened the pull request.

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -31,27 +31,21 @@ extern "C" {
+#include "irods/get_file_descriptor_info.h"
+#include "irods/replica_close.h"
+#include "irods/thread_pool.hpp"
+#include "irods/filesystem.hpp"
```
All of these should be using angle brackets instead of quotation marks, e.g.
```c++
#include <irods/filesystem.hpp>
```
Same for the other iRODS includes throughout this PR.

JustinKyleJames replied:
Should we change all of the irods includes or just the ones in this PR?

I generally prefer on this issue to be consistent with the legacy code rather than having a mix of quotes and angle brackets for irods includes but also hesitate to touch legacy code that doesn't require touching.

JustinKyleJames replied:
These were all change to angle brackets.

alanking commented on CMakeLists.txt:
```diff
@@ -23,7 +23,7 @@
 
-find_package(IRODS 4.2.11 EXACT REQUIRED CONFIG)
+find_package(IRODS 4.3.0 EXACT REQUIRED CONFIG)
 set(IRODS_CLIENT_REVISION "1")
```
Should we knock this back down to 0?

trel replied:
indeed.

trel replied:
diff still says 1... maybe you just haven't pushed yet?

JustinKyleJames replied:
Not sure what happened but this is done now.

JustinKyleJames commented:
I changed the code to remove the irods::globus::base64_encode/decode methods and used the irods::base64_encode/decode.  This will require the 4.3.1 release of iRODS before this can be released.

### More detail

1. We were picking up the wrong base64 methods so a local version was implemented for the checksum operations.  This was done in 4.2.x.
2. This wasn't sufficient for 4.3 because the clientLogin() code also calls the base64 methods.  Since this connector is not calling those directly, the workaround from above would not work with 4.3.
3. In iRODS 4.3.1, we changed the base64 methods to be in the irods namespace.  This allowed clientLogin() to work.  In addition, I was able to use these methods in the checksum operations so I was able to remove my local version of the base64 methods.

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -781,8 +780,11 @@ iRODS_connect_and_login(
         }
     }
 
+    printf("*****  iRODS: %s connected now logging in.\n", call_context_for_logging.c_str());
```
Should this be removed?

JustinKyleJames replied:
Good catch.  I put that in to debug the connection issue.  Will remove.

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -913,6 +915,7 @@ globus_l_gfs_iRODS_start(
                 }
                 iRODS_handle->user = strdup(token);
             }
+            iRODS_handle->first_write_done = false;
```
Should we initialize this value to `false` outside of the conditional block? The memory allocated on 841/843 doesn't seem to be set, so it may be a good idea to just zero out the struct, in which case this value would be `false` anyway. Thoughts?

alanking replied:
Oh, y'know what... if it's no a C-struct, that might not be a good idea. In any case, perhaps this value should be initialized/set outside of the conditional block.

JustinKyleJames replied:
I moved the initialization to outside the conditional block although I believe that first strtok() will always return a non-null pointer.

I am hesitant to zero initialize the entire struct because there are mutexes in it and I am not sure if that will cause problems.

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -1740,7 +1923,7 @@ globus_l_gfs_iRODS_recv(
     // start N threads to write to file
     int number_of_irods_write_threads = iRODS_handle->number_of_irods_read_write_threads;
-    irods::thread_pool threads{number_of_irods_write_threads};
+    irods::thread_pool threads{number_of_irods_write_threads-1};
```
This is -1 because the main thread is also writing, but separate from the `thread_pool`? Worth adding a comment to that effect?

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -1891,171 +2074,24 @@ globus_l_gfs_iRODS_recv(
 
     // start number_of_irods_write_threads threads to read from circular buffer and write to iRODS
-    for (int thr_id = 0; thr_id < number_of_irods_write_threads; ++thr_id)
+    for (int thr_id = 1; thr_id < number_of_irods_write_threads; ++thr_id)
```
Maybe worth a comment explaining why we start at 1 (because `thr_id` 0 is also a writer, but not part of the `thread_pool`?)

JustinKyleJames replied:
done

alanking commented on DSI/globus_gridftp_server_iRODS.cpp:
```diff
@@ -1674,6 +1677,186 @@ globus_l_gfs_iRODS_command(
 
 }
 
+void run_writer_thread(
```
The word `run` in this function name makes me think that this is spawning threads when in fact it is the task being posted to the `thread_pool`. Does it make sense to choose a name that communicates this? Maybe it's just me.

JustinKyleJames replied:
Maybe execute_writer_thread_operation?

JustinKyleJames replied:
Done

alanking reviewed (commented):
Nice work on the `base64_(en|de)code` stuff. Just a few more things I noticed this time around...

JustinKyleJames commented:
Can we take another look at this and see if it is ready to merge?

alanking reviewed (commented):
Yep, looks good to me. Let's squash it down to the 1-3 commits you want at the end

JustinKyleJames commented:
Squashed to 2 commits.

alanking commented:
I think... you might have rebased on top of 4-2-stable? Was that on purpose? I see 4 commits.

trel commented:
same, i see four.

trel commented:
now i see two.
